pkg/audit/verifier.go

```diff
@@ -489,24 +517,53 @@ func (verifier *Verifier) GetShare(ctx context.Context, limit *pb.AddressedOrder
 	}, nil
 }
 
+// RemoveFailedPieces removes lost pieces from a pointer
+func (verifier *Verifier) RemoveFailedPieces(ctx context.Context, path string, pointer *pb.Pointer, failedNodes storj.NodeIDList) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	if len(path) == 0 {
+		return nil
+	}
+
+	remoteSegment := pointer.GetRemote()
+	newRemotePieces := remoteSegment.RemotePieces[:0]
+	for _, piece := range remoteSegment.RemotePieces {
+		isFailed := false
+		for _, failedNode := range failedNodes {
+			if piece.NodeId == failedNode {
+				isFailed = true
+				break
+			}
+		}
+		if !isFailed {
+			newRemotePieces = append(newRemotePieces, piece)
+		}
+	}
+	remoteSegment.RemotePieces = newRemotePieces
+
+	// Update the segment pointer in the metainfo
+	//TODO:  update in a safe manner - https://storjlabs.atlassian.net/browse/V3-2088
+	return verifier.metainfo.Put(ctx, path, pointer)
+}
+
 // checkIfSegmentDeleted checks if stripe's pointer has been deleted since stripe was selected.
-func (verifier *Verifier) checkIfSegmentDeleted(ctx context.Context, stripe *Stripe) (err error) {
+func (verifier *Verifier) checkIfSegmentDeleted(ctx context.Context, segmentPath string, oldPointer *pb.Pointer) (newPointer *pb.Pointer, err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	pointer, err := verifier.metainfo.Get(ctx, stripe.SegmentPath)
+	newPointer, err = verifier.metainfo.Get(ctx, segmentPath)
 	if err != nil {
 		if storage.ErrKeyNotFound.Has(err) {
-			return ErrSegmentDeleted.New(stripe.SegmentPath)
+			return newPointer, ErrSegmentDeleted.New(segmentPath)
 		}
-		return err
+		return newPointer, err
 	}
-
-	if pointer.GetCreationDate().GetSeconds() != stripe.Segment.GetCreationDate().GetSeconds() ||
-		pointer.GetCreationDate().GetNanos() != stripe.Segment.GetCreationDate().GetNanos() {
-		return ErrSegmentDeleted.New(stripe.SegmentPath)
+	if oldPointer != nil {
+		if newPointer.GetCreationDate().GetSeconds() != oldPointer.GetCreationDate().GetSeconds() ||
+			newPointer.GetCreationDate().GetNanos() != oldPointer.GetCreationDate().GetNanos() {
+			return newPointer, ErrSegmentDeleted.New(segmentPath)
```

Comment:
Why don't we return `nil` instead of `newPointer` on error?

Reply:
I'm not sure if matters, but fixing